### ISSUE TITLE
Rule request: useless generator call

### ISSUE THREAD

**Issue by xavdid** [astral-sh/ruff#19018]
### Summary

Because generators don't execute until they're consumed, I had a small bug today where I expected a function call to throw and it didn't.

```py
import pytest

def my_gen():
    yield 1
    yield 2
    
    raise ValueError()

def test_raises(): # test fails incorrectly
    with pytest.raises(ValueError):
        my_gen()
```

Though everything looks correct (and passes all the lints I could find), `test_raises` fails because the expression doesn't actually raise.

To produce working code, I'd need to consume the generator:

```diff
def test_raises(): # test fails incorrectly
    with pytest.raises(ValueError):
-       my_gen()
+       list(my_gen())
```

And then it passes as expected.

It would be great if it could warn me that the generator call won't do anything. The original code should fail the lint, but assigning it to a variable or using the result in another function (like `list`) should pass.

Thanks!

**Comment by ntBre** [comment]
This sounds reasonable to me, but I'm open to other thoughts.

Somewhat related: https://github.com/astral-sh/ruff/issues/9833. This rule could benefit from better type inference too, but we could potentially handle the single-file case for now.

**Comment by MichaReiser** [comment]
The single file is tricky because it only works if the called functions annotate their return types. Which isn't the case for the example provided by @xavdid. I'm not sure how useful this rule is without return type inference
